README.pod
=begin html

<p>
<a href="https://travis-ci.org/nwellnhof/perl-commonmark?branch=master"><img src="https://travis-ci.org/nwellnhof/perl-commonmark.svg?branch=master" alt="Travis CI status"></a>
<a href="https://ci.appveyor.com/project/nwellnhof/perl-commonmark/branch/master"><img src="https://ci.appveyor.com/api/projects/status/rtjlp1r75mjf5m09/branch/master?svg=true" alt="AppVeyor status"/></a>
<a href="https://coveralls.io/r/nwellnhof/perl-commonmark?branch=master"><img src="https://coveralls.io/repos/nwellnhof/perl-commonmark/badge.svg?branch=master" alt="Coverage Status"></a>
<a href="http://badge.fury.io/pl/CommonMark"><img src="https://badge.fury.io/pl/CommonMark.svg" alt="CPAN version"></a>
</p>

=end html

=head1 Perl bindings for the CommonMark C library

This distribution is a wrapper around I<libcmark>, the official CommonMark C
library.

=head2 Installation of libcmark

Please note that the I<libcmark> API isn't stable yet. This version of the
Perl bindings is targeted at I<libcmark>
L<version 0.24.1|https://github.com/jgm/cmark/releases/tag/0.24.1>.
It's known to work with all releases between 0.21.0 and 0.24.1, but there's
no guarantee that it can be compiled with later versions. Also note that
upgrading libcmark may require recompilation of the Perl distribution.

    curl -LJO https://github.com/jgm/cmark/archive/0.24.1.tar.gz
    tar xzf cmark-0.24.1.tar.gz
    cd cmark-0.24.1
    make [INSTALL_PREFIX=/prefix]
    make test
    make install

See the I<libcmark> README for details.

=head2 Installation from a CPAN tarball

If I<libcmark> is in a standard location:

    perl Makefile.PL
    make
    make test
    make install

Otherwise:

    perl Makefile.PL \
        INC="-I/prefix/include" \
        LIBS="-L/prefix/lib -lcmark"
    make
    make test
    make install

See the documentation of I<ExtUtils::MakeMaker> for additional options.
The I<PERL_MM_OPT> environment variable is especially useful.

    export PERL_MM_OPT='INC="-I..." LIBS="-L... -lcmark"'

=head2 Build from a repository checkout 

This distribution uses I<Dist::Zilla> with the external plugins
I<MakeMaker::Awesome> and I<CopyFilesFromBuild>. You can build and test with
I<dzil>:

    dzil test
    dzil build

The files generated by I<Dist::Zilla> are included in the repository,
so you can use the standard build process as well.

=head2 Copyright

This software is copyright (C) by Nick Wellnhofer.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Perl module CommonMark

0.240100  2016-02-17
    - Support custom nodes (requires libcmark 0.23)
    - Export constants
    - Document more helper functions
    - Use new node naming conventions for helpers (old names are still
      supported).
    - Fix tests under libcmark 0.24.0.
      Thanks to A. Sinan Unur for the report.
    - Implement $node->replace

0.230002  2015-12-30
    - Fixes for libcmark 0.21.0

0.230001  2015-12-30
    - Define constants in XS code.
      Should add backward compatibility for libcmark 0.21.

0.230000  2015-12-29
    - Support and require libcmark 0.23
    - Improve and fix documentation
    - Fix "unescaped left brace" warning on Perl 5.22

0.210001  2015-07-28
    - Stop returning encoded UTF-8 from renderers
      BREAKING CHANGE: The renderers now return normal Perl strings.
      Use utf8::encode before printing the result to binary streams.
    - $node->render convenience method
    - Additional parser options for CommonMark->parse
    - Fix and improve check for libcmark

0.210000  2015-07-27
    - Require libcmark 0.21.0
    - Adjust for libcmark API changes
    - Add support for Latex renderer

0.180300  2015-04-02
    - Require libcmark 0.18.3
    - Add options parameter to markdown_to_html
    - Add support to render as Commonmark

0.1802  2015-03-16
    - Fix build with Perl <= 5.14.
    - Test with Travis CI.

0.1801  2015-03-11
    - Fix Parser->new to work correctly with libcmark 0.18.
      Also fixes compilation with libcmark 0.18.1.

0.1800  2015-03-05
    - Adjust to API changes in libcmark 0.18.
    - Require libcmark 0.18.
    - Add "parse" convenience function.
    - Fix build with old EUMM versions.

0.1700  2015-02-04
    - Check for libcmark with Devel::CheckLib.
    - Accessors for libcmark version.
    - Require libcmark 0.16.
    - Improve and fix documentation.

0.1601  2015-01-22
    - Initial release.



lib/CommonMark.pm
use 5.008;
use strict;
use warnings;

use XSLoader;

BEGIN {
    our $VERSION = '0.230002';
    XSLoader::load('CommonMark', $VERSION);
}

package CommonMark;

use Exporter 'import';
our %EXPORT_TAGS = (
    opt => [ qw(
        OPT_DEFAULT
        OPT_SOURCEPOS
        OPT_HARDBREAKS
        OPT_SAFE
        OPT_NORMALIZE
        OPT_VALIDATE_UTF8
        OPT_SMART
    ) ],
    node => [ qw(
        NODE_NONE
        NODE_DOCUMENT
        NODE_BLOCK_QUOTE
        NODE_LIST
        NODE_ITEM
        NODE_CODE_BLOCK
        NODE_HTML
        NODE_PARAGRAPH
        NODE_HEADER
        NODE_HRULE
        NODE_TEXT
        NODE_SOFTBREAK
        NODE_LINEBREAK
        NODE_CODE
        NODE_INLINE_HTML
        NODE_EMPH
        NODE_STRONG
        NODE_LINK
        NODE_IMAGE
    ),
    # libcmark 0.23
    qw(
        NODE_CUSTOM_BLOCK
        NODE_CUSTOM_INLINE
        NODE_HTML_BLOCK
        NODE_HEADING
        NODE_THEMATIC_BREAK
        NODE_HTML_INLINE
    ) ],
    list => [ qw(
        NO_LIST
        BULLET_LIST
        ORDERED_LIST
    ) ],
    delim => [ qw(
        NO_DELIM
        PERIOD_DELIM
        PAREN_DELIM
    ) ],
    event => [ qw(
        EVENT_NONE
        EVENT_DONE
        EVENT_ENTER
        EVENT_EXIT
    ) ],
);
Exporter::export_ok_tags();
push(@{ $EXPORT_TAGS{all} }, @{ $EXPORT_TAGS{$_} }) for keys %EXPORT_TAGS;

my @option_map = (
    sourcepos     => OPT_SOURCEPOS,
    hardbreaks    => OPT_HARDBREAKS,
    safe          => OPT_SAFE,
    normalize     => OPT_NORMALIZE,
    validate_utf8 => OPT_VALIDATE_UTF8,
    smart         => OPT_SMART,
);

sub parse {
    my ($class, %opts) = @_;

    my ($string, $file) = @opts{ qw(string file) };

    my $parser_opts = _extract_opts(\%opts);

    my $doc;
    if (defined($string)) {
        die("can't provide both string and file")
            if defined($file);
        $doc = $class->parse_document($string, $parser_opts);
    }
    elsif (defined($file)) {
        $doc = $class->parse_file($file, $parser_opts);
    }
    else {
        die("must provide either string or file");
    }

    return $doc;
}

sub _extract_opts {
    my $hash = shift;

    my $int = 0;

    for (my $i = 0; $i < @option_map; $i += 2) {
        my ($name, $val) = @option_map[$i,$i+1];
        $int |= $val if $hash->{$name};
    }

    return $int;
}

sub create_document {
    my (undef, %opts) = @_;
    my $node = CommonMark::Node->new(NODE_DOCUMENT);
    return _add_children($node, \%opts);
}

sub create_block_quote {
    my (undef, %opts) = @_;
    my $node = CommonMark::Node->new(NODE_BLOCK_QUOTE);
    return _add_children($node, \%opts);
}

sub create_list {
    my (undef, %opts) = @_;

    my $node = CommonMark::Node->new(NODE_LIST);

    my ($type, $delim, $start, $tight)
        = @opts{ qw(type delim start tight) };

    $node->set_list_type($type)   if defined($type);
    $node->set_list_delim($delim) if defined($delim);
    $node->set_list_start($start) if defined($start);
    $node->set_list_tight($tight) if defined($tight);

    return _add_children($node, \%opts);
}

sub create_item {
    my (undef, %opts) = @_;
    my $node = CommonMark::Node->new(NODE_ITEM);
    return _add_children($node, \%opts);
}

sub create_code_block {
    my (undef, %opts) = @_;

    my $node = CommonMark::Node->new(NODE_CODE_BLOCK);

    my $fence_info = $opts{fence_info};
    $node->set_fence_info($fence_info) if defined($fence_info);

    return _add_literal($node, \%opts);
}

sub create_html_block {
    my (undef, %opts) = @_;
    my $node = CommonMark::Node->new(NODE_HTML);
    return _add_literal($node, \%opts);
}

sub create_html { &create_html_block; }

sub create_paragraph {
    my (undef, %opts) = @_;
    my $node = CommonMark::Node->new(NODE_PARAGRAPH);
    return _add_children_or_text($node, \%opts);
}

sub create_heading {
    my (undef, %opts) = @_;

    my $node = CommonMark::Node->new(NODE_HEADER);

    my $level = $opts{level};
    $node->set_header_level($level) if defined($level);

    return _add_children_or_text($node, \%opts);
}

sub create_header { &create_heading; }

sub create_thematic_break {
    return CommonMark::Node->new(NODE_HRULE);
}

sub create_hrule { &create_thematic_break; }

sub create_text {
    my (undef, %opts) = @_;
    my $node = CommonMark::Node->new(NODE_TEXT);
    return _add_literal($node, \%opts);
}

sub create_softbreak {
    return CommonMark::Node->new(NODE_SOFTBREAK);
}

sub create_linebreak {
    return CommonMark::Node->new(NODE_LINEBREAK);
}

sub create_code {
    my (undef, %opts) = @_;
    my $node = CommonMark::Node->new(NODE_CODE);
    return _add_literal($node, \%opts);
}

sub create_html_inline {
    my (undef, %opts) = @_;
    my $node = CommonMark::Node->new(NODE_INLINE_HTML);
    return _add_literal($node, \%opts);
}

sub create_inline_html { &create_html_inline; }

sub create_emph {
    my (undef, %opts) = @_;
    my $node = CommonMark::Node->new(NODE_EMPH);
    return _add_children_or_text($node, \%opts);
}

sub create_strong {
    my (undef, %opts) = @_;
    my $node = CommonMark::Node->new(NODE_STRONG);
    return _add_children_or_text($node, \%opts);
}

sub create_link {
    my (undef, %opts) = @_;
    my $node = CommonMark::Node->new(NODE_LINK);
    return _add_link_opts($node, \%opts);
}

sub create_image {
    my (undef, %opts) = @_;
    my $node = CommonMark::Node->new(NODE_IMAGE);
    return _add_link_opts($node, \%opts);
}

sub create_custom_block {
    my (undef, %opts) = @_;
    my $node = CommonMark::Node->new(NODE_CUSTOM_BLOCK);
    return _add_custom_opts($node, \%opts);
}

sub create_custom_inline {
    my (undef, %opts) = @_;
    my $node = CommonMark::Node->new(NODE_CUSTOM_INLINE);
    return _add_custom_opts($node, \%opts);
}

sub _add_children {
    my ($node, $opts) = @_;

    my $children = $opts->{children};
    if (defined($children)) {
        for my $child (@$children) {
            $node->append_child($child);
        }
    }

    return $node;
}

sub _add_literal {
    my ($node, $opts) = @_;

    my $literal = $opts->{literal};
    $node->set_literal($literal) if defined($literal);

    return $node;
}

sub _add_children_or_text {
    my ($node, $opts) = @_;

    my $children = $opts->{children};
    my $literal  = $opts->{text};

    if (defined($children)) {
        die("can't set both children and text")
            if defined($literal);
        return _add_children($node, $opts);
    }

    if (defined($literal)) {
        my $text = __PACKAGE__->create_text(literal => $literal);
        $node->append_child($text);
    }

    return $node;
}

sub _add_link_opts {
    my ($node, $opts) = @_;

    my $url   = $opts->{url};
    my $title = $opts->{title};

    $node->set_url($url)     if defined($url);
    $node->set_title($title) if defined($title);

    return _add_children_or_text($node, $opts);
}

sub _add_custom_opts {
    my ($node, $opts) = @_;

    my $on_enter = $opts->{on_enter};
    my $on_exit  = $opts->{on_exit};

    $node->set_on_enter($on_enter) if defined($on_enter);
    $node->set_on_exit($on_exit)   if defined($on_exit);

    return _add_children_or_text($node, $opts);
}

package CommonMark::Node;

sub render {
    my ($self, %opts) = @_;

    my $format = $opts{format};
    die("must provide format")
        if !defined($format);
    my $method = "render_$format";

    my $render_opts = CommonMark::_extract_opts(\%opts);

    if ($format =~ /^(html|xml)\z/) {
        return $self->$method($render_opts);
    }
    if ($format =~ /^(commonmark|latex|man)\z/) {
        return $self->$method($render_opts, $opts{width});
    }

    die("invalid format '$format'");
}

1;



Makefile.PL
# This Makefile.PL for CommonMark was generated by
# Dist::Zilla::Plugin::MakeMaker::Awesome 0.38.
# Don't edit it but the dist.ini and plugins used to construct it.

use strict;
use warnings;

use 5.008;
use ExtUtils::MakeMaker;

use Devel::CheckLib;
{
local @ARGV;
unshift(@ARGV, ExtUtils::MakeMaker::_shellwords($ENV{PERL_MM_OPT} || ''))
if $ExtUtils::MakeMaker::VERSION >= 6.73_08;
check_lib(
header   => 'cmark.h',
function => 'return CMARK_VERSION >= 0x001500 ? 0 : 1;',
) or warn('libcmark 0.21.0 or higher not found'), exit;
}

my %WriteMakefileArgs = (
  "ABSTRACT" => "Interface to the CommonMark C library",
  "AUTHOR" => "Nick Wellnhofer <wellnhofer\@aevum.de>",
  "CONFIGURE_REQUIRES" => {
    "Devel::CheckLib" => 0,
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "CommonMark",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.008",
  "NAME" => "CommonMark",
  "PREREQ_PM" => {},
  "TEST_REQUIRES" => {
    "Test::More" => 0
  },
  "VERSION" => "0.240100",
  "test" => {
    "TESTS" => "t/*.t"
  }
);

%WriteMakefileArgs = (
    %WriteMakefileArgs,
    LIBS => '-lcmark',
);

my %FallbackPrereqs = (
  "Test::More" => 0
);

unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
